1. Importing libraries
2. Define dataframes
2a. Review statistics of df
2b. Practice solving mixed type columns
3. Find missing values
4. Address missing values
5. Find duplicate values
6. Address duplicate values
    Run consistency checks

In [2]:
#import libraries
import pandas as pd
import numpy as np
import os

2. Defining dataframes

In [3]:
# shortcut the main file to path
path = r"C:\Users\dodge\24-06-2023 Instacart Basket Analysis"

In [4]:
# define the dataframe for orders
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [5]:
#define the dataframe for products
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
#look at the statistics of the orders df
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [8]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) !=df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [11]:
# Change mixed type column to string
df_test['mix'] = df_test['mix'].astype('str')

3. Finding Missing Values

In [12]:
#Looking for missing values in products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Create subset of product_name null observations
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

4.  Addressing Missing Values

In [15]:
df_prods.shape

(49693, 5)

In [16]:
# Create a new dataframe removing missing values in product_name
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

5. Duplicate Values

In [18]:
#looking for full duplicates within the products dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

6. Addressing Duplicates

In [20]:
# Identify shape before removing dups
df_prods_clean.shape

(49677, 5)

In [21]:
#Removing the identified duplicates
df_prods_clean_no_dupe = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dupe.shape

(49672, 5)

# Data Consistency on df_ords dataframe

4.5.2  Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [23]:
#Review the stats on the orders dataframe
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Q2-The max order number is only 100, which compared to the order id seems small.

4.5.3  Check for mixed-type data in your df_ords dataframe.

In [24]:
#Checking the orders dataframe for mixed-type data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) !=df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

Q3-No mixed values found in the orders dataframe

In [25]:
#Confirming mixed types by looking at data types of each observation
df_ords.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [26]:
#Overwriting the dataframe to remove eval_set that was not overwritten previously
df_ords = df_ords.drop(columns = ['eval_set'])

In [27]:
#Rechecking data types
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

4.5.4  If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
None found.

4.5.5  Run a check for missing values in your df_ords dataframe.

In [28]:
#Looking for missing values in the orders dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Q5-The only missing cells are found in days_since_prior_order and there are a lot of them--over 200,000.  If this is a manual-order entry field, this could have been missed in order entry.  Or, this could represent a users first ever order, and thus there would be no prior order.  We can run some checks to determine if this represents user's first order.

4.5.6  Address the missing values using an appropriate method.

In [29]:
# Create subset of product_name null observations
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [30]:
df_nan_ords

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
11        2168274        2             1          2                 11   
26        1374495        3             1          1                 14   
39        3343014        4             1          6                 11   
45        2717275        5             1          3                 12   
...           ...      ...           ...        ...                ...   
3420930    969311   206205             1          4                 12   
3420934   3189322   206206             1          3                 18   
3421002   2166133   206207             1          6                 19   
3421019   2227043   206208             1          1                 15   
3421069   3154581   206209             1          3                 11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

Q6-The number of rows where days_since_prior_order is NaN is equal to 206,209 rows and all shown reflect order_number '1' as the customer's first order making these missing values approrpriate.  We could create a new column as a flag to designate first order for a user.  This could be a string value of "First Order" and "Repeat".  This could be an important data point for our customer--number of new customers versus repeat.

4.5.7 Run a check for duplicate values in your df_ords data.

In [31]:
#Looking for duplicates in the orders dataframe
df_dups_ords = df_ords[df_ords.duplicated()]

In [32]:
df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found!

4.5.8 Address the duplicates using an appropriate method.  NONE FOUND!

Last Step-exporting the new csv files

In [33]:
#Export the products dataframe into the prepared path
df_prods_clean_no_dupe.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [34]:
#Export the orders dataframe into the prepared path
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [35]:
df_ords.shape

(3421083, 6)